### PR TITLE
If ResourceCountCheckTask throws an exception the scheduled task is not going to run again until the management servers are restarted.

### PR DESCRIPTION
**Pull request by mprokopchuk**
Added try/catch to log the issue, continue processing of the next resources and running again according to the schedule.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [X] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [X] Minor
- [ ] Trivial

### How Has This Been Tested?

Tested manually

### CODE REVIEW

File: server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java [hunk 1]
```diff
@@ -1156,8 +1156,24 @@ public ResourceCountCheckTask() {
...
+                domains = new ArrayList<>();
+            }
+            // try/catch task, otherwise it won't be rescheduled in case of exception
+            try {
+                accounts = _accountDao.findActiveAccountsForDomain(Domain.ROOT_DOMAIN);
+            } catch (Exception e) {
+                s_logger.warn("Resource counters recalculation periodic task failed, unable to fetch active accounts for domain " + Domain.ROOT_DOMAIN, e);
+                // initialize accounts as empty list to do best effort recalculation
+                accounts = new ArrayList<>();
+            }
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java:1176]: does it make sense to keep running if one of these retrievals fail (for this pass I mean)?
I like the logging but i think a little further restructuring makes sense; two separate methods, one for domains and one for accounts for instance.
the recursion at lines 1197 and 1204 looks quite strange, if at all it works I would suggest a renaming for readability's sake.

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by shwstppr — APPROVED** [review]
LGTM

**Comment by shwstppr** [comment]
@blueorangutan package

File: server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java (hunk 1 above)
Comment (mprokopchuk) [reply server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java:1176]: It is Shadowing in Java nested classes: https://docs.oracle.com/javase/tutorial/java/javaOO/nested.html#shadowing

Comment (DaanHoogland) [reply server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java:1176]: yeah, well this doesn't make the code very readable. If there isn't a very strong design argument to do it this way, please get rid of it.

**Review by DaanHoogland — CHANGES_REQUESTED** [review]

**Comment by mprokopchuk** [comment]
@DaanHoogland renamed methods to recalculateAccountResourceCountInContext and recalculateDomainResourceCountInContext

File: server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java (hunk 1 above)
Comment (mlsorensen) [reply server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java:1176]:
```
does it make sense to keep running if one of these retrievals fail (for this pass I mean)?
```
Probably, I guess? At least we get the error logged now and if there is a problem with that specific account or domain's data it won't effect processing of the others.

Comment (mlsorensen) [reply server/src/main/java/com/cloud/resourcelimit/ResourceLimitManagerImpl.java:1176]: > does it make sense to keep running if one of these retrievals fail (for this pass I mean)?

Probably, I guess? At least we get the error logged now and if there is a problem with that specific account or domain's data it won't affect processing of the others.

> I like the logging but i think a little further restructuring makes sense; two separate methods, one for domains and one for accounts for instance.
> the recursion at lines 1197 and 1204 looks quite strange, if at all it works I would suggest a renaming for readability's sake.

I think changing the method names to be distinct has been addressed now, can we resolve the conversation?

**Comment by mlsorensen** [comment]
@blueorangutan package

**Comment by mlsorensen** [comment]
@blueorangutan test

**Comment by mlsorensen** [comment]
Let's try again...

@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by mlsorensen** [comment]
Errors look unrelated to this code change, they're at the KVM hypervisor layer (maybe due to an earlier commit, seems the live volume migration feature is unsupported in whatever host):

```
org.libvirt.LibvirtException: internal error: unable to execute QEMU command \\\'migrate\\\': this feature or command is not currently supported
```

lgtm

**Review by mlsorensen — APPROVED** [review]

**Pull request merged**